### PR TITLE
CNN Architecture & Train

### PR DESCRIPTION
**Pull request by abbyortego**
Created CNN
Trained CNN
Many things happened...im sorry

### CODE REVIEW

File: data-science/cnn/train.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+# TRAIN MODEL
+for e in range(num_epochs):
+    print(f'Epoch {e+1} ----------------------------------------------------------')
+    train(train_dataloader, model, loss_fn, optimizer)
+    test(test_dataloader, model, loss_fn)
+    torch.save(model.state_dict(), 'data-science/cnn/model_av/model_a_v2.pt')
+    print("Saved Tenative PyTorch Model State")
+
+print('All Done, you can cry over the accuracy now 	。゜゜(´Ｏ`) ゜゜。')
```
Comment (DeannaShort) [inline data-science/cnn/train.py:83]: :')

File: data-science/testin-er-out.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+if img.shape[2] == 4:
+    img = cv2.resize(cv2.cvtColor(img, cv2.COLOR_BGRA2BGR), [421, 614])
+else:
+    img = cv2.resize(cv2.cvtColor(img, cv2.COLOR_RGB2BGR), [421, 614])
+
+# convert to pytorch tensor (via PIL Image) and normalize
+tensor = from_PIL(Image.fromarray(np.uint8(img))).to(torch.float32)
+card = tensor / torch.max(tensor)
+card = card.unsqueeze(0)
+# print(card.shape)
```
Comment (DeannaShort) [inline data-science/testin-er-out.py:35]: for lines 19, 20 and 35, if this commented code isnt needed, we should remove it

**Review by DeannaShort — APPROVED** [review]

File: data-science/cnn/model_lv/card_classifier_l.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            nn.Conv2d(in_channels=3, out_channels=16, kernel_size=(3,3), stride = 3, padding=0),
+            nn.ReLU(),
+            nn.MaxPool2d((3,3), stride = 3) # may want to get rid of
+        )
+        
+        # self.conv2 = nn.Sequential(
+        #     nn.Conv2d(in_channels=32, out_channels=16, kernel_size=(3,3), stride = 3, padding=0),
+        #     nn.ReLU(),
+        #     nn.MaxPool2d((3,3), stride = 3)
+        # )
```
Comment (gbrad777) [inline data-science/cnn/model_lv/card_classifier_l.py:34]: this guy needed? if not should we take it out?

File: data-science/cnn/pytorch_tst.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+# c2_relu = torch.nn.functional.relu(c2_output, inplace=True)
+# # print(c1_relu)
+# img_cnv = to_PIL(c2_relu[7])
+# img_cnv.show()
+
+# # MX POOL
+# c2_mxpool_output = mxpool(c2_relu)
+# # print(c1_mxpool_output)
+# img_cnv = to_PIL(c2_mxpool_output[7])
+# img_cnv.show()
```
Comment (gbrad777) [inline data-science/cnn/pytorch_tst.py:80]: same thing for these guys, idk what they do but if you dont need em anymore maybe they go bye bye?

File: data-science/cnn/train.py (hunk 1 above)
Comment (gbrad777) [reply data-science/cnn/train.py:83]: 🥺

File: data-science/testin-er-out.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+card_classify = CardClassifier()
+card_classify.load_state_dict(torch.load('data-science/cnn/model_av/model_a_v2.pt'))
+card_classify.eval()
+
+from_PIL = transform.Compose([transform.PILToTensor()]) 
+
+# # get and load up card from URL
+# resp = requests.get('https://images.ygoprodeck.com/images/cards/88204302.jpg', headers = {'X-Api-Key': IDK_API_KEY, 'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36'})
+# img = np.array(Image.open(BytesIO(resp.content)))
```
Comment (gbrad777) [inline data-science/testin-er-out.py:20]: same dealio as before... do we need these guys👉👈?

File: data-science/testin-er-out.py (hunk 2 above)
Comment (abbyortego) [reply data-science/testin-er-out.py:35]: this is just a test file for the nueral net. it works for both cards from a url and cards from a directory which is the commented section. this file probably won't exist in the long run once querying starts but this code will be useful in the near future for a short period of time.

File: data-science/cnn/model_lv/card_classifier_l.py (hunk 3 above)
Comment (abbyortego) [reply data-science/cnn/model_lv/card_classifier_l.py:34]: yea this isnt needed anymore

File: data-science/cnn/pytorch_tst.py (hunk 4 above)
Comment (abbyortego) [reply data-science/cnn/pytorch_tst.py:80]: it's a test file for when building out CNN versions. it won't exist in the final product but must exist if we want to ever make improvements to our architecture in between now and then

File: data-science/testin-er-out.py (hunk 5 above)
Comment (abbyortego) [reply data-science/testin-er-out.py:20]: yes, they're for getting the cards by URL. it's just a test file and will get deleted at some point soon but is needed for now.

File: data-science/cnn/pytorch_tst.py (hunk 4 above)
Comment (gbrad777) [reply data-science/cnn/pytorch_tst.py:80]: coolio 👍

File: data-science/testin-er-out.py (hunk 5 above)
Comment (gbrad777) [reply data-science/testin-er-out.py:20]: okedoke 👍

**Review by gbrad777 — APPROVED** [review]

**Review by DeannaShort — APPROVED** [review]

**Pull request merged**